src/StringTools/RopeBuilder.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace StringTools
+{
+    /// <summary>
+    /// A StringBuilder replacement that keeps a list of <see cref="ReadOnlyMemory{T}"/> spans making up the intermediate string rather
+    /// than a copy of its characters. This has positive impact on both memory (no need to allocate space for intermediate strings) and
+    /// time (no need to copy characters to intermediate strings).
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="ToString"/> method tries to intern the resulting string without even allocating it if it's already interned.
+    /// Use <see cref="StringTools.GetRopeBuilder"/> to take advantage of pooling to eliminate allocation overhead of this class.
+    /// </remarks>
+    public class RopeBuilder : IDisposable
```

Comment:
As I understand it, Ropes is a well understood structure that is a balanced tree. I would avoid using the term for a list of spans. SpanCollection? SpanOfCharList?

Reply:
Renamed to `SpanBasedStringBuilder`. The name is a bit longer but I want to make it clear that it is supposed to be functionally analogous to `StringBuilder`.